tools/chpl-language-server/src/chpl-language-server.py

```diff
@@ -1185,29 +1268,89 @@ def unpack_call_hierarchy_item(
 
         return (fi, fn, instantiation)
 
+    def _all_end_marker_patterns(self) -> Dict[str, Tuple]:
+        return {
+            "loop": (chapel.Loop,),
+            "decl": (chapel.NamedDecl,),
+            "block": (
+                chapel.On,
+                chapel.Cobegin,
+                chapel.Begin,
+                chapel.Defer,
+                chapel.Serial,
+                chapel.Sync,
+                chapel.Local,
+                chapel.Manage,
+            ),
+        }
+
+    def _make_end_marker_patterns(self) -> Set[chapel.AstNode]:
+        to_match = set()
+        patterns = self._all_end_marker_patterns()
+        if "all" in self.end_markers:
+            for pat in patterns.values():
+                to_match.update(pat)
+        elif "none" not in self.end_markers:
+            for marker in self.end_markers:
+                to_match.update(patterns[marker])
+        return to_match
+
+    def get_end_marker(
+        self, node: chapel.AstNode, file_lines: List[str]
+    ) -> Optional[InlayHint]:
+        """
+        Get inlay hints that mark the end of a significant block of code. This
+        function assumes the feature is enabled for the node type, but does
+        check if the block is less than some threshold.
+        """
+        end_loc = location_to_range(node.location()).end
+        header_loc: Optional[chapel.Location] = None
+        goto_loc: Optional[chapel.Location] = None
+        if isinstance(node, chapel.Loop) and node.block_style() != "implicit":
+            header_loc = node.header_location()
+        elif isinstance(node, chapel.NamedDecl):
+            header_loc = node.header_location()
+            goto_loc = node.name_location()
+        elif isinstance(
+            node, self._all_end_marker_patterns().get("block", ())
+        ) and (
+            not isinstance(node, chapel.SimpleBlockLike)
+            or node.block_style() != "implicit"
+        ):
```

Comment:
This is awkward, you have some kind of two-way coupling between patterns (in all_end_marker_patterns) and this. Seems like what you want is two things: a pattern and a function that, given a node that matches this pattern, returns the header and goto locations. I have some ideas on how to clean this up, but maybe you as the author have better ones?

Reply:
I have some ideas. One is to store a lambda/function in the patterns table. I think this is the simplest approach. 

A slightly more verbose way would be to store real objects from a dataclass in this map. 

Do you have other approaches in mind?

Reply:
I was thinking lambda. You'd swiftly run into type errors I think (hence my offline comment about Python needing dependent types). Another approach is subclasses.

Reply:
Subclassing would have been my verbose approach, but I prefer the lambda approach